File: library/src/main/java/com/github/lany192/decoration/GridItemDecoration.java
package com.github.lany192.decoration;


import androidx.annotation.ColorInt;
import androidx.recyclerview.widget.GridLayoutManager;

/**
 * 适用于GridLayoutManager
 */
public class GridItemDecoration extends ItemDecoration {
    private int spanCount;
    private int width = 1;
    /**
     * 默认透明色
     */
    @ColorInt
    private int color = 0x00000000;

    /**
     * 是否显示四周的边
     */
    private boolean isShowBorder = false;

    public GridItemDecoration() {

    }

    public GridItemDecoration(int spanCount, int width, int color) {
        this.spanCount = spanCount;
        checkSpanCount(spanCount);
        this.width = width;
        this.color = color;
    }

    public GridItemDecoration(GridLayoutManager manager) {
        this.spanCount = manager.getSpanCount();
        checkSpanCount(spanCount);
    }

    public GridItemDecoration setSpanCount(int spanCount) {
        this.spanCount = spanCount;
        checkSpanCount(spanCount);
        return this;
    }

    public GridItemDecoration setWidth(int width) {
        this.width = width;
        return this;
    }

    public GridItemDecoration setColor(@ColorInt int color) {
        this.color = color;
        return this;
    }

    public GridItemDecoration setShowBorder(boolean isShowBorder) {
        this.isShowBorder = isShowBorder;
        return this;
    }

    @Override
    public Divider getDivider(int position) {
        checkSpanCount(spanCount);
        int remainder = position % spanCount;
        Border border = new Border(color, width);
        if (isShowBorder) {//要显示四周的边
            if (remainder == 0) {
                if (position < spanCount) {
                    return new Divider()
                            .setTop(border)
                            .setLeft(border)
                            .setRight(border)
                            .setBottom(border);
                } else {
                    return new Divider()
                            .setLeft(border)
                            .setRight(border)
                            .setBottom(border);
                }
            } else {
                if (position < spanCount) {
                    return new Divider().setTop(border)
                            .setRight(border)
                            .setBottom(border);
                } else {
                    return new Divider().setRight(border).setBottom(border);
                }
            }
        } else {
            if (remainder == spanCount - 1) {
                //最后一个只显示bottom
                return new Divider().setBottom(border);
            } else {
                return new Divider().setRight(border).setBottom(border);
            }
        }
    }

    private void checkSpanCount(int spanCount) {
        if (spanCount < 2) {
            throw new RuntimeException("SpanCount minimum is 2");
        }
    }
}

File: library/src/main/java/com/github/lany192/decoration/GridItemDecoration.kt
package com.github.lany192.decoration

import androidx.annotation.ColorInt
import androidx.recyclerview.widget.GridLayoutManager

/**
 * 适用于GridLayoutManager
 */
class GridItemDecoration : ItemDecoration {
    private var spanCount = 0
    private var width = 1

    /**
     * 默认透明色
     */
    @ColorInt
    private var color = 0x00000000

    /**
     * 是否显示四周的边
     */
    private var isShowBorder = false

    constructor() {}
    constructor(spanCount: Int, width: Int, color: Int) {
        this.spanCount = spanCount
        checkSpanCount(spanCount)
        this.width = width
        this.color = color
    }

    constructor(manager: GridLayoutManager) {
        spanCount = manager.spanCount
        checkSpanCount(spanCount)
    }

    fun setSpanCount(spanCount: Int): GridItemDecoration {
        this.spanCount = spanCount
        checkSpanCount(spanCount)
        return this
    }

    fun setWidth(width: Int): GridItemDecoration {
        this.width = width
        return this
    }

    fun setColor(@ColorInt color: Int): GridItemDecoration {
        this.color = color
        return this
    }

    fun setShowBorder(isShowBorder: Boolean): GridItemDecoration {
        this.isShowBorder = isShowBorder
        return this
    }

    override fun getDivider(position: Int): Divider? {
        checkSpanCount(spanCount)
        val remainder = position % spanCount
        val border =
            Border(color, width.toFloat())
        return if (isShowBorder) { //要显示四周的边
            if (remainder == 0) {
                if (position < spanCount) {
                    Divider()
                        .setTop(border)
                        .setLeft(border)
                        .setRight(border)
                        .setBottom(border)
                } else {
                    Divider()
                        .setLeft(border)
                        .setRight(border)
                        .setBottom(border)
                }
            } else {
                if (position < spanCount) {
                    Divider().setTop(border)
                        .setRight(border)
                        .setBottom(border)
                } else {
                    Divider().setRight(border)
                        .setBottom(border)
                }
            }
        } else {
            if (remainder == spanCount - 1) {
                //最后一个只显示bottom
                Divider().setBottom(border)
            } else {
                Divider().setRight(border).setBottom(border)
            }
        }
    }

    private fun checkSpanCount(spanCount: Int) {
        if (spanCount < 2) {
            throw RuntimeException("SpanCount minimum is 2")
        }
    }
}